### DISCUSSION TITLE
자료구조 - 스택/큐/덱 - 서희&재찬

### DISCUSSION THREAD

**Discussion by LireEruel** [Shortudy-10th/tech-for-developer#51] [질의 응답]
- 큐, 스택, 덱이 각각 어떠한 구조를 가지는지 설명해 주세요.
- 스택, 큐, 덱을 프로그램상에서 구현한다면, 어떻게 구현할 수 있을까요?
- OS, 어플리케이션 등 다양한 프로그램에서 스택, 큐, 덱을 실제로 어떻게 활용하고 있나요? 그 예시에 대해 설명해주세요.
- 브라우저에서 스택과 큐는 언제 활용되고 있나요?
- 스택과 큐의 특징과 이 둘은 어디에 주로 사용되나요?

**Comment by LireEruel** [comment]
### ❓️큐, 스택, 덱이 각각 어떠한 구조를 가지는지 설명해 주세요.

**Reply by LireEruel** [reply]
스택(Stack): 스택은 LIFO(Last In First Out) 구조로, 마지막에 들어간 요소가 먼저 나오는 형태입니다. 즉, 가장 최근에 추가된 요소가 먼저 제거됩니다.
큐(Queue): 큐는 FIFO(First In First Out) 구조로, 먼저 들어간 요소가 먼저 나오는 형태입니다.
덱(Deque): 덱은 양쪽 끝에서 삽입과 삭제가 모두 가능한 자료구조로, 스택과 큐의 특징을 모두 가지고 있습니다.

**Reply by LireEruel** [reply]
## ❓️ 스택 오버플로우와 스택 언더플로우는 무엇이며, 어떻게 방지할 수 있나요?

- 스택 오버플로우는 스택의 용량을 초과하여 데이터를 삽입하려는 경우 발생합니다. 스택 언더플로우는 스택이 비어 있는데 데이터를 꺼내려는 경우 발생합니다. 이를 방지하기 위해서는 스택의 크기를 적절히 관리하고, 삽입/삭제 연산 전에 스택의 상태를 확인하는 로직을 추가해야 합니다.

**Comment by LireEruel** [comment]
## ❓️스택, 큐, 덱을 프로그램 상에서 구현한다면, 어떻게 구현할 수 있을까요?

**Reply by kjc6735** [reply]
어레이리스트나 링크드리스트를 이용하여 구현할 수 있다.

**Comment by LireEruel** [comment]
## ❓️ OS, 어플리케이션 등 다양한 프로그램에서 스택, 큐, 덱을 실제로 어떻게 활용하고 있나요? 그 예시에 대해 설명해주세요.

**Reply by kjc6735** [reply]
스택의 예시로는 함수 콜 스택이 있다. 프로그램에서 함수를 호출할 때 스택에 호출 정보를 추가하고, 함수가 끝날 때 스택에서 제거한다.

큐 예시로는 스케쥴링이 있다. 여러가지 작업을 큐에 넣어두고 순서대로 처리한다.

덱의 예시로는 웹 브라우저에서의 탭 관리나, 음악 재생 목록가 있다.

**Comment by LireEruel** [comment]
## ❓️ 브라우저에서 스택과 큐는 언제 활용되고 있나요?

**Reply by LireEruel** [reply]
**스택 **
- 브라우저의 뒤로 가기 기능에서 페이지 이력을 관리
- JavaScript의 실행 컨텍스트 관리에 사용됩니다.
- 브라우저의 렌더링 엔진에서는 스택을 사용하여 HTML 태그의 계층 구조를 파싱하고, 큐를 사용하여 비동기 이벤트나 콜백을 관리합니다. 이를 통해 웹 페이지의 구조와 동작을 효율적으로 처리할 수 있습니다.
큐
- 비동기 작업의 실행 순서를 관리하는 이벤트 큐에서 사용됩니다.

**Comment by LireEruel** [comment]
## ❓️ 스택과 큐의 특징과 이 둘은 어디에 주로 사용되나요?

**Reply by LireEruel** [reply]
스택: 가장 나중에 들어온 데이터가 먼저 처리되어야 하는 경우에 사용됩니다. 함수 호출, 문법 검사, 연산자 우선순위 처리 등에 활용됩니다.
큐: 가장 먼저 들어온 데이터가 먼저 처리되어야 하는 경우에 사용됩니다. 데이터 순차 처리, 작업 스케줄링, 네트워크 통신 등에 활용됩니다.

**Reply by LireEruel** [reply]
❓️ : 큐와 스택을 조합하여 어떤 문제를 해결할 수 있을까요? 예를 들어 설명해주세요.
- 큐와 스택을 조합하여 이중 스택 큐를 구현할 수 있습니다. 두 개의 스택을 사용하여 큐의 동작을 에뮬레이션할 수 있으며, 이를 통해 다양한 자료 처리 문제를 해결할 수 있습니다.
